# An R shiny application for visualization of EndoC-bH1 and human islet genomic data

## Overview of App

EndoC-βH1 is emerging as a critical human beta cell model to study the genetic and environmental etiologies of beta cell (dys)function and diabetes. Comprehensive knowledge of its molecular landscape is lacking, yet required, for effective use of this model. Here, we report chromosomal (spectral karyotyping), genetic (genotyping), epigenomic (ChIP-seq, ATAC-seq), chromatin interaction (Hi-C, Pol2 ChIA-PET), and transcriptomic (RNA-seq, miRNA-seq) maps of EndoC-βH1. Analyses of these maps define known (e.g., PDX1, ISL1) and putative (e.g., PCSK1, mir-375) beta cell-specific transcriptional cis-regulatory networks and identify allelic effects on cis-regulatory element use. Importantly, comparison with maps generated in primary human islets/beta cells indicate preservation of chromatin looping, but also highlight chromosomal aberrations and fetal genomic signatures in EndoC-βH1. Together, these maps, and a web application we created for their exploration, provide important tools for the design of experiments to probe and manipulate the genetic programs governing beta cell identity and (dys)function in diabetes.

## App Availability

This app is free to use and hosted at: https://shinyapps.jax.org/endoc-islet-multi-omics/

## App Usage/Tutorial

For a detailed walk-through of how to use this application please see this document: https://github.com/nlawlor/Multiomic_Browser/blob/master/Data/EndoC_Islet_Omics_Shiny_App_Manual.docx

## Publication Link

To learn more about this study and app please read our publication: https://www.cell.com/cell-reports/fulltext/S2211-1247(18)32043-6

![](https://github.com/nlawlor/Multiomic_Browser/blob/master/Graphical.abstract.FINAL.jpg)
